### PR TITLE
Correctly html encode text

### PR DESCRIPTION
**Pull request by sniffdk**
Before returning the html string, referencing issue #6226

### CODE REVIEW

**Comment by sniffdk** [comment]
Sorry, that went a bit too fast, I've reverted and added a new method to replace the old one.

File: src/Umbraco.Web/HtmlStringUtilities.cs [hunk 1]
```diff
@@ -20,9 +20,25 @@ public sealed class HtmlStringUtilities
         /// </summary>
         /// <param name="text">The text.</param>
         /// <returns>The text with text line breaks replaced with HTML line breaks (<br/>)</returns>
+        [Obsolete("Doesn't html encode the text. Use ReplaceLineBreaks instead")]
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/HtmlStringUtilities.cs]:
```suggestion
        [Obsolete("This method doesn't HTML encode the text. Use ReplaceLineBreaks instead.")]
```

File: src/Umbraco.Web/HtmlStringUtilities.cs [hunk 2]
```diff
@@ -20,9 +20,25 @@ public sealed class HtmlStringUtilities
...
+        }
+        
+        /// <summary>
+        /// Html encodes the text and replaces text line breaks with HTML line breaks.
+        /// </summary>
+        /// <param name="text">The text.</param>
+        /// <returns>The html encoded text with text line breaks replaced with HTML line breaks (<br/>)</returns>
+        public HtmlString ReplaceLineBreaks(string text)
+        {
+            var value = HttpUtility.HtmlEncode(text ?? string.Empty)
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/HtmlStringUtilities.cs]: `HttpUtility.HtmlEncode()` and `HtmlString` happily accept `null` values, so no need to create empty strings. This also ensures the `Replace` calls are skipped when the text is `null` (by using the null-conditional operator):

```suggestion
            var value = HttpUtility.HtmlEncode(text)?
```

Comment (sniffdk) [reply src/Umbraco.Web/HtmlStringUtilities.cs]: Sure, makes sense 👍

File: src/Umbraco.Web/HtmlStringUtilities.cs [hunk 3]
```diff
@@ -20,9 +20,25 @@ public sealed class HtmlStringUtilities
         /// </summary>
         /// <param name="text">The text.</param>
         /// <returns>The text with text line breaks replaced with HTML line breaks (<br/>)</returns>
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/HtmlStringUtilities.cs]:
```suggestion
        /// <returns>The text with text line breaks replaced with HTML line breaks (<c>&lt;br /&gt;</c>).
```

File: src/Umbraco.Web/HtmlStringUtilities.cs [hunk 4]
```diff
@@ -20,9 +20,25 @@ public sealed class HtmlStringUtilities
...
+        [Obsolete("This method doesn't HTML encode the text. Use ReplaceLineBreaks instead.")]
         public HtmlString ReplaceLineBreaksForHtml(string text)
         {
-            return new HtmlString(text.Replace("\r\n", @"<br />").Replace("\n", @"<br />").Replace("\r", @"<br />"));
+            return new HtmlString(text.Replace("\r\n", @"<br />").Replace("\n", @"<br />").Replace("\r", @"<br />"));            
+        }
+        
+        /// <summary>
+        /// Html encodes the text and replaces text line breaks with HTML line breaks.
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/HtmlStringUtilities.cs]:
```suggestion
        /// HTML encodes the text and replaces text line breaks with HTML line breaks.
```

File: src/Umbraco.Web/HtmlStringUtilities.cs [hunk 5]
```diff
@@ -20,9 +20,25 @@ public sealed class HtmlStringUtilities
...
+        /// </summary>
+        /// <param name="text">The text.</param>
+        /// <returns>The html encoded text with text line breaks replaced with HTML line breaks (<br/>)</returns>
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/HtmlStringUtilities.cs]:
```suggestion
        /// <returns>The HTML encoded text with text line breaks replaced with HTML line breaks (<c>&lt;br /&gt;</c>).</returns>
```

File: src/Umbraco.Web/HtmlStringUtilities.cs [hunk 6]
```diff
@@ -20,9 +20,25 @@ public sealed class HtmlStringUtilities
...
+        public HtmlString ReplaceLineBreaks(string text)
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/HtmlStringUtilities.cs]:
```suggestion
        public IHtmlString ReplaceLineBreaks(string text)
```

Comment (sniffdk) [reply src/Umbraco.Web/HtmlStringUtilities.cs]: Yeah, we might as well, as that is what Umbraco uses in property value converters etc.

**Comment by nul800sebastiaan** [comment]
Thanks @sniffdk for updating!

I don't really like that we're repeating the `Replace` code now but I guess it's fine since we wouldn't ever want to change it anyway because that would be a breaking change.

I guess the `Truncate` method suffers from the same problem and we should probably audit all methods that return `IHtmlString` / `HtmlString`.

**Comment by sniffdk** [comment]
Also thanks to @ronaldbarendse 👍 
Not sure about the need to modify Truncate as I would assume that it expects html in the first place, considering the amount of work it does.

**Pull request merged**